### ISSUE TITLE
into_iter_on_ref autofix fails

### ISSUE THREAD

**Issue by SebastianSpeitel** [rust-lang/rust-clippy#15761]
### Summary

Trying to fix a `into_iter_on_ref` warningfails:
```
warning: failed to automatically apply fixes suggested by rustc to crate `netherite_protocol`

after fixes were automatically applied the compiler reported errors within these files:

  * netherite-protocol/src/lib.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0271]: expected `impl Iterator<Item = u8>` to be an iterator that yields `&_`, but it yields `u8`
    --> netherite-protocol/src/lib.rs:210:21
     |
210  |         self.iter().copied()
     |                     ^^^^^^ expected `&_`, found `u8`
     |
     = note: expected reference `&_`
                     found type `u8`
note: the method call chain might not have had the expected associated types
    --> netherite-protocol/src/lib.rs:210:14
     |
210  |         self.iter().copied()
     |         ---- ^^^^^^ `Iterator::Item` is `u8` here
     |         |
     |         this expression has type `&[u8; N]`
note: required by a bound in `std::iter::Iterator::copied`
    --> /usr/lib/rust/1.89.0/lib/rustlib/src/rust/library/core/src/iter/traits/iterator.rs:3419:32
     |
3417 |     fn copied<'a, T: 'a>(self) -> Copied<Self>
     |        ------ required by a bound in this associated function
3418 |     where
3419 |         Self: Sized + Iterator<Item = &'a T>,
     |                                ^^^^^^^^^^^^ required by this bound in `Iterator::copied`

error[E0271]: expected `impl Iterator<Item = u8>` to be an iterator that yields `&_`, but it yields `u8`
   --> netherite-protocol/src/lib.rs:209:23
    |
209 |     fn iter(&self) -> impl Iterator<Item = u8> {
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^ expected `&_`, found `u8`
210 |         self.iter().copied()
    |         -------------------- return type was inferred to be `std::iter::Copied<impl std::iter::Iterator<Item = u8>>` here
    |
    = note: expected reference `&_`
                    found type `u8`
    = note: required for `std::iter::Copied<impl std::iter::Iterator<Item = u8>>` to implement `std::iter::Iterator`

error[E0720]: cannot resolve opaque type
   --> netherite-protocol/src/lib.rs:209:23
    |
209 |     fn iter(&self) -> impl Iterator<Item = u8> {
    |                       ^^^^^^^^^^^^^^^^^^^^^^^^
```

Original diagnostics:
```
warning: this `.into_iter()` call is equivalent to `.iter()` and will not consume the `array`
   --> netherite-protocol/src/lib.rs:210:14
    |
210 |         self.into_iter().copied()
    |              ^^^^^^^^^ help: call directly: `iter`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#into_iter_on_ref
    = note: `#[warn(clippy::into_iter_on_ref)]` on by default
```

### Reproducer

I tried this code:

```rust
trait ByteArray {
    fn iter(&self) -> impl Iterator<Item = u8>;
}
impl<const N: usize> ByteArray for [u8; N] {
    fn iter(&self) -> impl Iterator<Item = u8> {
        self.into_iter().copied()
    }
}
```

```
cargo clippy --fix
```

I expected to see this happen:
Either no diagnostic, because it's not possible to fix or a fix that works.

Instead, this happened:
Clippy fails to fix the diagnostic

### Version

```text

```
